metric_providers/base.py

```diff
@@ -51,8 +51,18 @@ def check_system(self):
 
     # implemented as getter function and not direct access, so it can be overloaded
     # some child classes might not actually have _ps attribute set
+    #
+    # This function has to go through quite some hoops to read the stderr
+    # The preferred way to communicate with processes is through communicate()
+    # However this function ALWAYS waits for the process to terminate and it does not allow reading from processes
+    # in chunks while they are running. Thus we we cannot set encoding='UTF-8' in Popen and must decode here.
     def get_stderr(self):
-        return self._ps.stderr.read()
+        stderr_read = ''
+        if self._ps.stderr is not None:
+            stderr_read = self._ps.stderr.read()
+            if isinstance(stderr_read, bytes):
+                stderr_read = stderr_read.decode('utf-8')
+        return stderr_read
```

Comment:
If we always return the empty string we can remove quite a few checks for stderr being None